src/erl/typespec.rs

```diff
@@ -0,0 +1,263 @@
+*/
+pub fn type_(name: &str, _parameters: &[String], constructors: &[RecordConstructor]) -> Document {
+    let found_parameters = extract_params_from_constructors(constructors);
+
+    "-type "
+        .to_doc()
+        .append(name.to_snake_case())
+        .append(wrap_args(
+            found_parameters
+                .iter()
+                .map(|name| name.to_camel_case().to_doc()),
+        ))
+        .append(" :: ")
+        .append(concat(
+            constructors
+                .iter()
+                .map(|constructor| {
+                    if constructor.args.is_empty() {
+                        // Basic single-atom constructor
+                        type_name_to_atom(&constructor.name).to_doc()
+                    } else if is_record(&constructor) {
+                        record_type(&constructor)
+                    } else {
+                        non_record_type(&constructor)
+                    }
+                })
+                .intersperse(delim(" |")),
+        ))
+        .append(".")
+}
+
+fn record_type(constructor: &RecordConstructor) -> Document {
+    let name = constructor.name.to_snake_case();
+    let contents = type_record_definition(&constructor.args);
+
+    "#".to_doc().append(name).append(contents)
+}
+
+fn non_record_type(constructor: &RecordConstructor) -> Document {
+    constructor
+        .name
+        .to_snake_case()
+        .to_doc()
+        .append(delim(","))
+        .append(concat(
+            constructor
+                .args
+                .iter()
+                .map(|(_, ast, ..)| type_ast_name(ast, None))
+                .intersperse(delim(",")),
+        ))
+        .surround("{", "}")
+}
+
+fn is_record(constructor: &RecordConstructor) -> bool {
+    for (label, ..) in constructor.args.iter() {
+        match label {
+            Some(_) => return true,
+            None => return false,
+        }
+    }
+
+    false
+}
+
+fn type_record_definition(args: &[(Option<String>, TypeAst, SrcSpan)]) -> Document {
+    concat(
+        args.iter()
+            .map(|(label, ast, ..)| {
+                label
+                    .clone()
+                    .unwrap_or("".to_string())
+                    .to_doc()
+                    .append(" :: ")
+                    .append(type_ast_name(&ast, None))
+            })
+            .intersperse(", ".to_doc()),
+    )
+    .surround("{", "}")
+}
+
+// Generates the spec for a regular Gleam function
+pub fn fun_spec(name: &str, args: &[TypedArg], return_type: &Type) -> Document {
+    "-spec "
+        .to_doc()
+        .append(atom(name.to_string()))
+        .append(spec_args(args))
+        .append(" -> ")
+        .append(typ_to_erl(return_type))
+        .append(".")
+}
+
+fn spec_args(args: &[TypedArg]) -> Document {
+    wrap_args(args.iter().map(|a| typ_to_erl(&*a.typ)))
+}
+
+fn typ_to_erl(typ: &Type) -> Document {
+    match typ {
+        Type::App { name, .. } => name.to_snake_case().to_doc(),
```

Comment:
We will need to namespace this type if it comes from another module

Reply:
I've attempted to do this now though I don't know how good an approach it is.